api/dataset_test.go

```diff
@@ -492,50 +758,170 @@ func TestPostVersionReturnsError(t *testing.T) {
+				return nil, internalError
 			},
 		}
+		mockedDataStore.GetEdition(bson.M{"links.dataset.id": "123", "edition": "2017"})
 
-		api := CreateDatasetAPI(secretKey, mux.NewRouter(), DataStore{Backend: mockedDataStore})
+		api := CreateDatasetAPI(host, secretKey, mux.NewRouter(), DataStore{Backend: mockedDataStore})
 		api.router.ServeHTTP(w, r)
 		So(w.Code, ShouldEqual, http.StatusInternalServerError)
-		So(len(mockedDataStore.UpsertVersionCalls()), ShouldEqual, 1)
+		So(len(mockedDataStore.GetEditionCalls()), ShouldEqual, 2)
 	})
 
-	Convey("When the api cannot connect to datastore to update dataset resource return an internal server error", t, func() {
+	Convey("When the api cannot connect to datastore to update version resource return an internal server error", t, func() {
 		var b string
-		b = "{\"edition\":\"2017\",\"state\":\"published\",\"license\":\"ONS\",\"release_date\":\"2017-04-04\",\"version\":\"1\"}"
-		r, err := http.NewRequest("POST", "http://localhost:22000/datasets/123/editions/2017/versions/1", bytes.NewBufferString(b))
+		b = versionPayload
+		r, err := http.NewRequest("POST", "http://localhost:22000/datasets/123/editions/2017/versions", bytes.NewBufferString(b))
+		r.Header.Add("internal_token", "coffee")
 		So(err, ShouldBeNil)
 		w := httptest.NewRecorder()
 		mockedDataStore := &backendtest.BackendMock{
-			UpsertVersionFunc: func(string, interface{}) error {
-				return nil
+			GetEditionFunc: func(interface{}) (*models.Edition, error) {
+				return &models.Edition{}, nil
 			},
-			UpsertDatasetFunc: func(string, interface{}) error {
+			GetNextVersionFunc: func(string, string) (int, error) {
+				return 1, nil
+			},
+			UpsertVersionFunc: func(string, interface{}) error {
 				return internalError
 			},
 		}
 
+		mockedDataStore.GetEdition(bson.M{"links.dataset.id": "123", "edition": "2017"})
+		mockedDataStore.GetNextVersion("123", "2017")
+
 		update := bson.M{
 			"$set": bson.M{
-				"state": "published",
+				"state":         "published",
+				"collection_id": "12345",
 			},
 			"$setOnInsert": bson.M{
 				"updated_at": time.Now(),
 			},
 		}
 		mockedDataStore.UpsertVersion("1", update)
 
-		api := CreateDatasetAPI(secretKey, mux.NewRouter(), DataStore{Backend: mockedDataStore})
+		api := CreateDatasetAPI(host, secretKey, mux.NewRouter(), DataStore{Backend: mockedDataStore})
 		api.router.ServeHTTP(w, r)
 		So(w.Code, ShouldEqual, http.StatusInternalServerError)
+		So(len(mockedDataStore.GetEditionCalls()), ShouldEqual, 2)
+		So(len(mockedDataStore.GetNextVersionCalls()), ShouldEqual, 2)
 		So(len(mockedDataStore.UpsertVersionCalls()), ShouldEqual, 2)
-		So(len(mockedDataStore.UpsertDatasetCalls()), ShouldEqual, 1)
 	})
+
+	Convey("When the request does not contain a valid internal token return status unauthorised", t, func() {
+		var b string
+		b = "{\"edition\":\"2017\",\"state\":\"created\",\"license\":\"ONS\",\"release_date\":\"2017-04-04\",\"version\":\"1\"}"
```

Comment:
Use backticks to write JSON. It is seen as a raw string literal.

Reply:
Thanks